FWCore/Framework/src/PathsAndConsumesOfModules.cc

```diff
@@ -288,196 +352,345 @@ namespace edm {
+            }
+          }
+        }
+      }
+    } catch (CircularDependencyException const&) {
+      //the last element in stackTrace must appear somewhere earlier in stackTrace
+      std::ostringstream oStr;
+
+      unsigned int lastIndex = stackTrace.front();
+      bool firstSkipped = false;
+      for (auto id : stackTrace) {
+        if (firstSkipped) {
+          oStr << "  module '" << moduleIndexToNames[lastIndex] << "' depends on " << moduleIndexToNames[id] << "\n";
+        } else {
+          firstSkipped = true;
+        }
+        lastIndex = id;
+      }
+      throw edm::Exception(edm::errors::ScheduleExecutionFailure, "Unrunnable schedule\n")
+          << "Circular module dependency found in configuration\n"
+          << oStr.str();
+    }
 
-    //determine the path dependencies
-    std::vector<std::string> pathNames = iPnC.paths();
-    const unsigned int kFirstEndPathIndex = pathNames.size();
+    auto pathName = [&](PathStatus const& iP) {
+      if (iP.endPath_) {
+        return iPnC.endPaths()[iP.index_];
+      }
+      return iPnC.paths()[iP.index_];
+    };
 
-    const std::string kPathEnded("@PathEnded");
-    const std::string kEndPathStart("@EndPathStart");
+    //The program would deadlock
+    if (not madeForwardProgress) {
+      std::ostringstream oStr;
+      auto modIndex = std::numeric_limits<unsigned int>::max();
+      unsigned int presentPath;
+      for (auto itP = statusOfPaths.begin(); itP != statusOfPaths.end(); ++itP) {
+        auto const& p = *itP;
+        if (p.activeModuleSlot_ == p.nModules_) {
+          continue;
+        }
+        //this path is stuck
+        modIndex = p.modulesOnPath_[p.activeModuleSlot_];
+        presentPath = itP - statusOfPaths.begin();
+        break;
+      }
+      //NOTE the following should always be true as at least 1 path should be stuc.
+      // I've added the condition just to be paranoid.
+      if (modIndex != std::numeric_limits<unsigned int>::max()) {
+        struct ProgressInfo {
+          ProgressInfo(unsigned int iMod, unsigned int iPath, bool iPreceeds = false)
+              : moduleIndex_(iMod), pathIndex_(iPath), preceeds_(iPreceeds) {}
 
-    //The finished processing depends on all paths and end paths
-    const std::string kFinishedProcessing("@FinishedProcessing");
-    const unsigned int kFinishedProcessingIndex{0};
-    moduleIndexToNames.insert(std::make_pair(kFinishedProcessingIndex, kFinishedProcessing));
+          ProgressInfo(unsigned int iMod) : moduleIndex_(iMod), pathIndex_{}, preceeds_(false) {}
 
-    pathNames.insert(pathNames.end(), iPnC.endPaths().begin(), iPnC.endPaths().end());
-    std::vector<std::vector<unsigned int>> pathIndexToModuleIndexOrder(pathNames.size());
-    {
-      for (unsigned int pathIndex = 0; pathIndex != pathNames.size(); ++pathIndex) {
-        std::set<unsigned int> alreadySeenIndex;
+          unsigned int moduleIndex_ = std::numeric_limits<unsigned int>::max();
```

Comment:
I think this initialization value "std::numeric_limits<unsigned int>::max()" is unnecessary. It's harmless, but all the constructors initialize moduleIndex_.